### ISSUE TITLE
There's no ID for the socket

### ISSUE THREAD

**Issue by francislavoie** [simonhamp/echo#2]
https://github.com/simonhamp/echo/blob/ratchet-connector/src/connector/ratchet-connector.ts#L28

If you compare with the socket.io implementation, socket.io returns an `id` property on the socket which is a unique ID (just some random string, not sure how they build it). Echo depends on it and sends it in the `X-Socket-Id` header to the server. The ratchet implementation probably needs some kind of replacement for that somehow.

**Comment by simonhamp** [comment]
Interesting. I haven’t seen a need for this in my tests but it looks like a good approach.

Has the lack of this caused any issues for you?

**Comment by francislavoie** [comment]
Not yet, but I noticed it while digging through the code. I couldn't get laravel-zmq broadcasting working yet (I don't think you wrote up an actual example for that anywhere yet?)

Thanks a bunch for your various libs/forks, it helped a lot!

**Comment by simonhamp** [comment]
Glad they’re proving useful 👍 

I found the ZMQ bit quite tricky too. It’s not super clear (IMHO) how you should set up your various clients with ZMQ through PHP. I think the trick that made things work in my earliest prototype (IIRC) was to set the socket type in my ZMQ config to PUSH on the ‘server’ side and PULL on the ‘client’ side (I.e. in Ratchet) - I’m pretty sure the default is PUB/SUB and I couldn’t get that to play ball for some reason.

I am working on a complete server implementation of Ratchet+ZMQ that should just be a drop-in package to get a fully working WS server running within Laravel. That will have a working example when it’s ready.

I’ll cycle back here with references once I have something concrete.

**Comment by francislavoie** [comment]
I actually tried logging right before the connection happens and nothing
shows up, so I don't think for me the connection is the issue. How are you
setting up the events and how do you broadcast? Making my own push socket
worked fine though, so I know zmq itself isn't the issue but rather
something with the laravel broadcasting.

**Comment by simonhamp** [comment]
I’m on holiday right now and don’t have my laptop with me so I can’t check. I know my implementation wasn’t fully working last time I checked, but I’m sure some of my code is online already...

If you take a look at what I have so far in [simonhamp/laravel-echo-ratchet-server](/simonhamp/laravel-echo-ratchet-server), it might give you some ideas.

If you don’t mind waiting for a week or so, I’ll get back onto development of that package when I return home and will be able to provide a more useful response.

**Comment by francislavoie** [comment]
Sounds good. I'm on holidays too but I tried to get some work done :P

I did look at that repo of yours but there isn't anything in the pertaining
to the actual broadcasting. It just has the ws server and the config files.

Thanks!

**Comment by simonhamp** [comment]
Ok that’s good. I think there may be some wires crossed somewhere. When you refer to broadcasting, I’m thinking about what actually handles the broadcasting of events from Laravel - the Broadcasting service.

For our needs here, broadcasting all happens in ZMQ thanks to the package that my echo Ratchet server implementation requires: the ZMQ Broadcast driver package I adopted (and collaborate on).

Do you mean some other part of the pipeline?

**Comment by francislavoie** [comment]
I mean that I do `broadcast(new SomeEvent());` but it doesn't seem to get sent through laravel-zmq as far as I can tell. I don't have a good grasp of those internals. I did configure the broadcast driver and set it to zmq, the service provider does get called though (dropped in some vardumps). I'm not at a PC either right now so no rush at all btw.

**Comment by simonhamp** [comment]
Ok, perhaps when we’re both back in front of our computers we can do a screen share call and work through it together?

**Comment by francislavoie** [comment]
So I got around to trying some more stuff out, I ultimately found that it was a misconfiguration on my part, but partly because of mistakes in the docs: https://github.com/pelim/laravel-zmq/pull/4

This issue still stands though, a socketId is still needed for proper functionality of the broadcasting flows. One feature Laravel provides is broadcasting to all subscribers _except_ the one who published it. This is implemented by skipping the subscriber with the matching socket ID.

**Comment by simonhamp** [comment]
Well spotted in the docs issue. Sorry about that. Good confirmation on this Socket ID header too. I’ll work something up next week.

Thanks for persisting with these libs and helping improve them. It’s really appreciated!

**Comment by francislavoie** [comment]
Looks like socket.io uses https://github.com/faeldt/base64id btw

https://github.com/socketio/engine.io/blob/b1fa020675d905905b6f07d0a6b5acff775c4be5/lib/server.js#L281

**Comment by simonhamp** [comment]
@francislavoie I've implemented a basic UUIDv4 ID generator to create socket IDs. This seems to be working now, but if you get a chance to play with it I'd appreciate your feedback as my use-case isn't quite the same as yours yet.

**Comment by francislavoie** [comment]
Awesome! Great timing, I was about to look into doing this myself soon :P
I'll give it a shot on Monday (or sooner if I get the inclination to do some work on the weekend).

**Comment by francislavoie** [comment]
Can confirm that HTTP requests with Axios do set the `X-Socket-Id` header with the latest build.

See here https://laravel.com/docs/5.5/broadcasting#broadcasting-events, the configuration section mentions how it's expected to work.

The idea is that `toOthers()` gets called, which reads the socket ID from the current request's header if it exists. Then Echo will filter out any message that has the same socket ID as itself (I think), which gets passed along in the payload of the message if it has the `InteractsWithSockets` trait. I didn't test with `toOthers()` yet, but I think this is a step forwards in properly supporting the laravel flows with ratchet.

**Comment by simonhamp** [comment]
Agreed, this should work, but I haven't tested it either yet. I'll leave this open until we've had chance to prove if it works.
